mlflow/sklearn/__init__.py

```diff
@@ -1200,8 +1247,10 @@ def fit_mlflow(original, self, *args, **kwargs):
         _log_pretraining_metadata(autologging_client, self, *args, **kwargs)
         params_logging_future = autologging_client.flush(synchronous=False)
         fit_output = original(self, *args, **kwargs)
-        _log_posttraining_metadata(autologging_client, self, *args, **kwargs)
-        autologging_client.flush(synchronous=True)
+        # params of xgboost sklearn models are logged in train() in mlflow.xgboost.autolog()
+        if flavor_name == FLAVOR_NAME:
+            _log_posttraining_metadata(autologging_client, self, *args, **kwargs)
+            autologging_client.flush(synchronous=True)
```

Comment:
@jwyyy Instead of special casing XGBoost logic in `fit_mlflow`, can we define a new method called `fit_mlflow_xgboost` that just calls `original(self, *args, **kwargs)` and then logs `self` using `mlflow.xgboost.log_model()`? This will also allow us to revert changes to XGBoost autologging's `train()` method, since we can control how the model gets logged here.

We can then add a parameter to `patched_fit` (https://github.com/mlflow/mlflow/blob/6e4b64bd3b7beadbdb243f6921d9ebedb10a1dbd/mlflow/sklearn/__init__.py#L1372) to specify either `fit_mlflow` (for sklearn models) or `fit_mlflow_xgboost` (for xgboost sklearn models). Perhaps we can call this parameter `fit_fn`.

Let me know if you have questions here!

Reply:
@dbczumar Thank you for your suggestion! There is a small issue on model logging in the `train()` method when calling `fit()`. The `fit()` method calls `train()` internally and assigns a Booster object to the internal `_Booster` in a XGBoost sklearn model [see [L1331](https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/sklearn.py#L1331)]. The current `train()` in `mlflow.xgboost.autolog()` logs models before returning the `model` object, which means (1) the logged model is a Booster object; (2) we cannot log XGBoost sklearn models before assigning the trained Booster to `_Booster`. The changes in `mlflow.xgboost.autolog()` try to log sklearn models directly. I think we definitely can log sklearn models in `fit_mlflow_xgboost()` but it is extra work. Because models are logged when calling `train()`, and calling `fit_mlflow_xgboost()` just logs new information to replace old ones. However, I also think adopting your suggestion makes the code logic easier to read. Please let me know which solution sounds better to you. Thank you!

Reply:
@jwyyy Ah, thanks for letting me know! Can we decompose `train()` into two methods - one for parameter, metric, and non-model artifact logging, and one for artifact logging? We can then use the former method to patch `xgboost.sklearn.train()`.

Reply:
Sounds good to me! I will make some adjustments.